mlir/lib/Dialect/Rock/Transforms/GemmOutputSwizzle.cpp

```diff
@@ -0,0 +1,571 @@
+      return failure();
+    }
+    int64_t blockSize = maybeBlockSize.value().getValue().getSExtValue();
+    int64_t gridSize = maybeGridSize.value().getValue().getSExtValue();
+    bool useIndexDiffs = true;
+    bool forceUnroll = true;
+    int64_t ldsRequiredBytes = mPerBlock * nPerBlock * getByteWidth(destType);
+
+    // Decide register vectorization.
+    constexpr int64_t dimensionM = 1;
+    constexpr int64_t dimensionN = 2;
+    int64_t dataPerThread = (nPerBlock * mPerBlock) / blockSize;
+    if ((nPerBlock * mPerBlock) % blockSize != 0) {
+      return failure();
+    }
+
+    VectorizationResult mVectorRes =
+        getMaxVectorization(matC, dimensionM, /*inputDimLen=*/
+                            dataPerThread, matC.getDefiningOp());
+    int64_t mVectorLen = mVectorRes.max;
+    VectorizationResult nVectorRes =
+        getMaxVectorization(matC, dimensionN, /*inputDimLen=*/
+                            dataPerThread, matC.getDefiningOp());
+    int64_t nVectorLen = nVectorRes.max;
+    int64_t dim = (mVectorLen > nVectorLen) ? dimensionM : dimensionN;
+
+    // Get current workitem ID.
+    auto tid = b.create<WorkitemIdOp>(loc, b.getIndexType());
+
+    // Allocate LDS for output.
+    auto workgroupMemoryAddressSpace = b.getAttr<gpu::AddressSpaceAttr>(
+        gpu::GPUDialect::getWorkgroupAddressSpace());
+    auto ldsMemRefOutputType =
+        MemRefType::get({ldsRequiredBytes}, b.getI8Type(), AffineMap{},
+                        workgroupMemoryAddressSpace);
+    auto ldsBufferOutput = b.create<GpuAllocOp>(loc, ldsMemRefOutputType);
+    auto typedBuffer = viewBufferAs(b, ldsBufferOutput, destType);
+
+    // Convert from raw -> mPerBlock, nPerBlock
+    TopDownTMBuilder mnToRaw(b, {"gemmM", "gemmN"}, {mPerBlock, nPerBlock});
+    if (dim == dimensionN) {
+      mnToRaw.unmerge("flatten", 0, {"gemmM", "gemmN"}, {mPerBlock, nPerBlock});
+    } else {
+      mnToRaw.unmerge("flatten", 0, {"gemmN", "gemmM"}, {nPerBlock, mPerBlock});
+    }
+    auto mnToRawAttr = mnToRaw.get();
+
+    SmallVector<Attribute> transformMNToRawAttrs;
+    transformMNToRawAttrs.push_back(mnToRawAttr);
+    ArrayAttr transformMNToRaw = b.getArrayAttr(transformMNToRawAttrs);
+
+    auto ldsBufferMNToRaw = transform(b, typedBuffer, transformMNToRaw);
+
+    // LDS barrier.
+    // This barrier is needed because we reuse A and B buffers
+    b.create<LDSBarrierOp>(loc);
+
+    // Store C results to LDS.
+    b.create<ThreadwiseWriteAllOp>(loc, convertedC, ldsBufferMNToRaw,
+                                   /*extraViews=*/idToLDS,
+                                   /*extraIndices=*/ValueRange{tid},
+                                   op.getFeatures(), StoreMethod::Set,
+                                   /*forceUnroll=*/forceUnroll,
+                                   /*useIndexDiffs=*/useIndexDiffs);
+
+    // Load from LDS to registers.
+    int64_t elementsWrittenPerThread = 128 / destType.getIntOrFloatBitWidth();
```

Comment:
There is an underlying assumption this always succeed for dwordx4 vectorization.
I d feel safer to use maxVecLen instead here.